quipucords/scanner/normalizer.py

```diff
@@ -0,0 +1,228 @@
+        :param server_id: server_id of quipucords server that collected these
+        raw facts.
+        """
+        self.raw_facts = raw_facts
+        self._metadata = {}
+        self.server_id = server_id
+
+    @property
+    def facts(self):
+        """Return normalized facts."""
+        if not hasattr(self, "_normalized"):
+            raise AssertionError(
+                "You must call '.normalize()' before acessing `.facts`."
+            )
+        return self._normalized
+
+    @property
+    def metadata(self):
+        """Return metadata for normalized facts."""
+        if not hasattr(self, "_normalized"):
+            raise AssertionError(
+                "You must call '.normalize()' before acessing `.metadata`."
+            )
+        return self._metadata
+
+    def normalize(self):
+        """Normalize and validate raw facts.
+
+        :return: dict of normalized facts
+        """
+        if hasattr(self, "_normalized"):
+            return self._normalized
+        self._normalized = {}
+        for fact_name in self.fields.keys():
+            norm_result = self._normalize_fact(fact_name)
+            self._normalized[fact_name] = norm_result.value
+            self._metadata[fact_name] = {
+                "raw_fact_keys": norm_result.raw_fact_keys,
+                "source_type": self.source_type,
+                "server_id": self.server_id,
+                "has_error": norm_result.has_error,
+            }
+        return self._normalized
+
+    def _normalize_fact(self, fact_name) -> "NormalizedResult":
+        """
+        Normalize given fact.
+
+        Blind exceptions are used to ensure issues with normalization or validation
+        functions don't affect other facts.
+
+        Returns the normalized fact and a list a raw facts used for normalization.
+        """
+        field: FactMapper = self.fields[fact_name]
+        kwargs = {dep: self._normalized[dep] for dep in field.dependencies}
+        if len(field.raw_fact_keys) == 1:
+            raw_fact_key = field.raw_fact_keys[0]
+            raw_value = self.raw_facts.get(raw_fact_key)
+            args = (raw_value,)
+        else:
+            args = ()
+            kwargs.update(
+                **{raw: self.raw_facts.get(raw) for raw in field.raw_fact_keys}
+            )
+        try:
+            norm_result = field.normalizer_func(*args, **kwargs)
+        except Exception:
+            logger.exception("Unexpected error during '%s' normalization.", fact_name)
+            return NormalizedResult(None, has_error=True, raw_fact_keys=None)
+        if not isinstance(norm_result, NormalizedResult):
+            raw_fact_keys = self._get_raw_fact_keys(fact_name)
+            norm_result = NormalizedResult(
+                value=norm_result, raw_fact_keys=raw_fact_keys
+            )
+        else:
+            self._check_raw_fact_keys_for_fact(fact_name, norm_result.raw_fact_keys)
```

Comment:
> This will happen when I claim that fact_foo depends on fact_bar, but fact_bar is not defined, correct? 

Yes

> So this is really developer-facing. And as a developer, I would prefer this thing to blow up loudly and scream at me that I've got my dependencies defined wrong.

This is a really good argument. Gonna change it here.